### ISSUE TITLE
[Errno 30] Read-only file system

### ISSUE THREAD

**Issue by Phylloxera** [tseemann/shovill#132]
Not sure how reproducible this will be, but here goes.
I'm running shovill 1.1 via a singularity container running ubuntu bionic. it works fine on our centos7 fat node (48 cores) but fails on our sge cluster when it gets to spades with the following error:

>[shovill] Will use spades 003014000 options: --isolate and --merged
[shovill] Using tempdir: /local/371702.1.work.q
.............................................................................................
 [spades] [Errno 30] Read-only file system: '/local'
[spades] Traceback (most recent call last):
[spades]   File "/usr/src/SPAdes-3.14.0-Linux/bin/spades.py", line 590, in main
[spades]     bin_home, ext_python_modules_home, python_modules_home)
[spades]   File "/usr/src/SPAdes-3.14.0-Linux/bin/spades.py", line 530, in build_pipeline
[spades]     ext_python_modules_home, python_modules_home)
[spades]   File "/usr/src/SPAdes-3.14.0-Linux/share/spades/spades_pipeline/stages/spades_stage.py", line 378, in add_to_pipeline
[spades]     dataset_data, log, bin_home, ext_python_modules_home, python_modules_home))
[spades]   File "/usr/src/SPAdes-3.14.0-Linux/share/spades/spades_pipeline/stages/spades_stage.py", line 266, in __init__
[spades]     self.generate_cfg(cfg, self.output_files)
[spades]   File "/usr/src/SPAdes-3.14.0-Linux/share/spades/spades_pipeline/stages/spades_stage.py", line 356, in generate_cfg
[spades]     self.cfg.tmp_dir = support.get_tmp_dir(prefix="spades_")
[spades]   File "/usr/src/SPAdes-3.14.0-Linux/share/spades/spades_pipeline/support.py", line 468, in get_tmp_dir
[spades]     os.makedirs(base_dir)
[spades]   File "/usr/lib/python2.7/os.py", line 150, in makedirs
[spades]     makedirs(head, mode)
[spades]   File "/usr/lib/python2.7/os.py", line 157, in makedirs
[spades]     mkdir(name, mode)
[spades] OSError: [Errno 30] Read-only file system: '/local'

This is particularly weird, because /local is writable... in fact, spades writes to it just fine when running shovill via conda (1.0.4), the relevant portion of the the shovill log:

>[shovill] Yes, it supports --pe1-m
[shovill] Using tempdir: /local/371599.1.work.q
...............................................................................................
 [spades] Other parameters:
[spades]   Dir for temp files: /local/371599.1.work.q

maybe i built spades wrong for use with -pe... or i need to downgrade from 2.14 to 2.13? ideas?

Thanks!

**Comment by kapsakcj** [comment]
Perhaps SPAdes is confusing `/local` inside the container vs `/local` outside the container (i.e. on the node it's running on)?

Have you tried setting the $TMPDIR manually as an environmental variable, or as an argument passed to SPAdes?

What's the full singularity command that you are passing to `qsub`? 

I had success running my Shovill 1.1.0 docker image via singularity on our cluster that runs UGE. It automatically used `/tmp` as the temporary directory
```bash
echo "singularity exec --cleanenv --no-home -B $(realpath .):/data ~/singularity-images/shovill.1.1.0.simg \
shovill --R1 /data/1.fastq.gz --R2 /data/2.fastq.gz --cpus 16 --ram 60 --outdir /data/shovill-out/" \
 | qsub -N shovill-sing-test -pe smp 16 -l h_vmem=82G -cwd -V -j y -o $(realpath .)/sge-sing-shovill/
```

**Comment by Phylloxera** [comment]
Hi @kapsakcj Thanks for giving this a go. I think you are probably right... and this is also something our sysadmin suggested. but is there a way to specify $TMPDIR in the batch file? I did some googling and couldn't find anything (at least that I could understand) I don't want to do it as an argument passed to spades since i'm using shovill as a component of a larger pipeline. Here is my current batch file"

> #!/bin/bash
#$ -N shovillhank2
#$ -q work.q
#$ -l h_rt=96:00:00
#$ -pe smp 40
#$ -V
#$ -o "stdout_shov3.txt"
#$ -e "stderr_shov3.txt"
#$ -wd /home/aaron.dickey
sh simplehank2.sh

I tried adding in:

> #$ -v $TMPDIR=/local

but it doesn't seem to have worked

**Comment by kapsakcj** [comment]
Try dropping the $ in your addition? Perhaps the bash script changes that to whatever TMPDIR is predefined as.
So:
> #$ -v TMPDIR=/local

If that doesn't do it, I'm thinking it has to be something passed to singularity. So something like `singularity exec -B /local:/local [...]` So that singularity knows to use `/local` on the node for TMPDIR as well as `/local` inside the container for writing temporary SPAdes files.

**Comment by tseemann** [comment]
Thanks for helping @kapsakcj

**Comment by Phylloxera** [comment]
Yes, thanks again so much @kapakcj for giving this a go and yes this is clearly a user_env issue.
Just wanted to give a follow-up:

> #$ -v TMPDIR=/local

inside the batch did not work. SPAdes inside the container is trying to revert to $TMPDIR in the location (specified by the SGE daemon?) yet doesn't have write access to that tmp-dir (because it is looking for that location inside the container?) I did find [this](https://wiki.rc.ucl.ac.uk/wiki/Example_Submission_Scripts), where they add a `cd $TMPDIR` command after all the batch specifications. This did actually work as expected for an individual shovill run on our cluster (I have no guess as to why but it worked both when placed in the batch and in the runscript). Unfortunately, it did not "scale-up" appropriately to my larger pipeline: Instead of shovill running on all 96 fq.gz pairs specified by the variable in my for loop, it ran on a smaller (apparently random) subset of them (maybe 60-or-so), weird.
I also tried including `--opts "--tmp-dir $TMPDIR"` in the shovill command but this didn't work either. I decided not to pursue the singularity -B specification since I'm seeking a solution that is robust to different server configurations. We'll see if there is any problem when I test on a SLURM cluster.
